lightning_examples/finetuning-scheduler/finetuning-scheduler.py

```diff
@@ -0,0 +1,712 @@
+
+# %%
+# Load our lightning module...
+lightning_module_kwargs = {
+    "model_name_or_path": "microsoft/deberta-v3-base",
+    "optimizer_init": optimizer_init,
+    "lr_scheduler_init": lr_scheduler_init,
+}
+model = RteBoolqModule(**lightning_module_kwargs, experiment_tag="fts_explicit")
+
+# %% [markdown]
+# ### Callback Configuration
+#
+# The only callback required to invoke the [FinetuningScheduler](https://finetuning-scheduler.readthedocs.io/en/stable/api/finetuning_scheduler.fts.html#finetuning_scheduler.fts.FinetuningScheduler) is the [FinetuningScheduler](https://finetuning-scheduler.readthedocs.io/en/stable/api/finetuning_scheduler.fts.html#finetuning_scheduler.fts.FinetuningScheduler) callback itself.
+# Default versions of [FTSCheckpoint](https://finetuning-scheduler.readthedocs.io/en/stable/api/finetuning_scheduler.fts_supporters.html#finetuning_scheduler.fts_supporters.FTSCheckpoint) and [FTSEarlyStopping](https://finetuning-scheduler.readthedocs.io/en/stable/api/finetuning_scheduler.fts_supporters.html#finetuning_scheduler.fts_supporters.FTSEarlyStopping)
+# (if not specifying ``epoch_only_transitions``) will be included ([as discussed above](#basic_usage)) if not provided
+# in the callbacks list. For demonstration purposes I'm including example configurations of all three callbacks below.
+
+# %%
+# let's save our callback configurations for the explicit scenario since we'll be reusing the same
+# configurations for the implicit and nofts_baseline scenarios (except the  config for the
+# FinetuningScheduler callback itself of course in the case of nofts_baseline)
+earlystopping_kwargs = {"monitor": "val_loss", "min_delta": 0.001, "patience": 2}
+checkpoint_kwargs = {"monitor": "val_loss", "save_top_k": 1}
+fts_kwargs = {"max_depth": 1}
+callbacks = [
+    FinetuningScheduler(ft_schedule=ft_schedule_name, **fts_kwargs),  # type: ignore # noqa
+    FTSEarlyStopping(**earlystopping_kwargs),  # type: ignore # noqa
+    FTSCheckpoint(**checkpoint_kwargs),  # type: ignore # noqa
+]
+
+# %%
+logger = TensorBoardLogger("lightning_logs", name="fts_explicit")
+# optionally start tensorboard and monitor progress graphically while viewing multi-phase finetuning specific training
+# logs in the cell output below by uncommenting the next 2 lines
+# # %load_ext tensorboard
+# # %tensorboard --logdir lightning_logs
+# disable progress bar by default to focus on multi-phase training logs. Set to True to re-enable if desired
+enable_progress_bar = False
+
+# %%
+
+
+def train() -> None:
+    trainer = pl.Trainer(
+        enable_progress_bar=enable_progress_bar,
+        max_epochs=100,
+        precision=16,
+        strategy="dp",
+        accelerator="gpu",
+        devices=1,
+        callbacks=callbacks,
+        logger=logger,
+    )
+    trainer.fit(model, datamodule=dm)
+
+
+if AVAIL_GPUS > 0:
+    train()
+else:
+    print("Given the multiple phases of finetuning demonstrated, this notebook is best used with a GPU")
```

Comment:
why is that?
just trying to understand how the gradual unfreezing is related to the accelerator being used.

Reply:
My language was somewhat unclear, thanks for pointing that out. I've updated the language to more clearly indicate that to avoid excessive runtime, I'm suggesting GPU acceleration. I figured the user can easily run with a CPU by commenting out the if guard if they want, but that it might be good to make the default behavior explicitly GPU so they don't get a nasty surprise waiting around forever on a CPU unexpectedly. :)